inject-dist/observers.js

```diff
@@ -1,19 +1,143 @@
-// Note: Using prefixed variables to avoid naming collisions in the global scope.
-// Map of features to the number of times they are accessed by the page.
-const httparchive_observers = {};
-// Preserve the initial state of any features whose getter methods will be overwritten.
-const httparchive_init = {
-  navigator: {
-    userAgent: navigator.userAgent
-  }
+// Note: Using prefixed variable to avoid naming collisions in the global scope.
+// This name must exactly match the one referenced in custom metrics.
+const httparchive_observers = {
+  call_stacks: {}
 };
+let httparchive_enable_observations = false;
+
+// Local scope.
+(() => {
+
+// Add the pathnames of any functions/properties you want to observe.
+const OBSERVERS = [
+  'navigator.__proto__.*',
+  'performance.__proto__.*',
+  'performance.timing.__proto__.*',
+  'Array.prototype.*',
+  'String.prototype.*',
+  'Object.prototype.*',
+  'CSSStyleDeclaration.prototype.*',
+  'document.featurePolicy',
+  'document.write',
+  'queueMicrotask',
+  'requestIdleCallback',
+  'scheduler.postTask'
+];
+
+const PROPERTIES_TO_TRACE = new Set([
+  'navigator.userAgent'
+]);
+
+
+function resolveObject(pathname) {
+  let obj = window;
+  let props = pathname.split('.');
+
+  while (props.length) {
+    if (!obj) {
+      return null;
+    }
+
+    const prop = props.shift();
+    obj = obj[prop];
+  }
+
+  return obj || null;
+}
+
+function getAllProperties(pathname, depth=1) {
+  if (!depth) {
+    return pathname;
+  }
+
+  const props = pathname.split('.');
+  const parentPathname = props.slice(0, -1).join('.');
+  const parentObj = resolveObject(parentPathname);
+
+  return Object.getOwnPropertyNames(parentObj).flatMap(prop => {
+    return getAllProperties(`${parentPathname}.${prop}`, depth - 1);
+  });
+}
+
+function initializeObserver(pathname) {
+  const props = pathname.split('.');
+  const prop = props.at(-1);
+  let parentPathname;
+  let original;
+
+  if (props.at(-2) == '__proto__') {
+    // Omit __proto__ for observation purposes.
+    parentPathname = props.slice(0, -2).join('.');
+  } else {
+    parentPathname = props.slice(0, -1).join('.');
+  }
+
+  pathname = `${parentPathname}.${prop}`;
+  const parentObj = resolveObject(parentPathname);
+
+  try {
+    original = parentObj[prop];
+  } catch (e) {
+    // The property is not accessible.
+    return;
+  }
+
+  try {
+    Object.defineProperty(parentObj, prop, {
+      configurable: true,
+      get: () => {
+        if (!httparchive_enable_observations) {
+          return original;
+        }
+
+        if (PROPERTIES_TO_TRACE.has(pathname)) {
+          // Construct a stack trace.
+          let stack;
+          try {
+            throw new Error();
+          } catch (e) {
+            stack = e.stack;
+          }
+          let stackCounter = httparchive_observers.call_stacks[pathname];
+          if (!stackCounter[stack]) {
+            stackCounter[stack] = 0;
+          }
+          stackCounter[stack]++;
```

Comment:
[no-plusplus](https://eslint.org/docs/rules/no-plusplus)

Reply:
I can't find it now, but there's a philosophy of "if the linter says it's ok, it's ok". Since the linter we currently use says it's ok, it's ok 😁

That said, if there are any specific improvements you think we can make to our code, let's fine tune the linter itself.